File: Seori/LeetCode 75/Graph_DFS/547. Number of Provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        
        def dfs(now: int) -> None:
            for next in range(n):
                if isConnected[now][next] == 1 and not visited[next]:
                    visited[next] = 1
                    dfs(next)

        n = len(isConnected)
        visited = [0] * n
        
        # 1번 도시부터 DFS 탐색을 실시한다. 다른 도시와 연결되지 않은 경우에만 is_province가 증가한다.
        is_province = 0
        for i in range(n):
            if not visited[i]:
                visited[i] = 1
                dfs(i)
                is_province += 1

        return is_province

File: Gombang/[LeetCode75] - [DFS]/547_Number_of_Provinces.cpp
class Solution {
public:

    unordered_map<int, vector<int>> graph;

    void dfs(int currCity, vector<bool>& visited)
    {
        if (visited[currCity])
            return;

        visited[currCity] = true;
        for (int adjCity : graph[currCity])
        {
            if (visited[adjCity])
                continue;

            dfs(adjCity, visited);
        }
    }

    int findCircleNum(vector<vector<int>>& isConnected) {

        // 어떤 도시와 어떤 도시가 연결되어 있는지 그래프로 표현.
        for (int i = 0; i < isConnected.size(); i++)
        {
            for (int j = 0; j < isConnected[i].size(); j++)
            {
                if (isConnected[i][j] == 1 && i != j)
                {
                    graph[i].push_back(j);
                    graph[j].push_back(i);
                }
            }
        }

        int cityCount = isConnected.size();
        vector<bool> visited(cityCount, false);

        int groupCount = 0;
        for (int i = 0; i < cityCount; i++)
        {
            if (visited[i])
                continue;

            groupCount++;
            dfs(i, visited);
        }

        return groupCount;
    }
};